<a href="https://colab.research.google.com/github/jiriyaporn/data_viz/blob/main/pandas102.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:

import os 

In [6]:
path = '/content/drive/My Drive/dataviz'

In [7]:
covid_file_path = os.path.join(path, 'pm-20-jan-2021.csv')
print(covid_file_path)

/content/drive/My Drive/dataviz/pm-20-jan-2021.csv


In [8]:
a = path + '/pm-20-jan-2021.csv'
print(a)

/content/drive/My Drive/dataviz/pm-20-jan-2021.csv


In [9]:
data_covid = pd.read_csv(covid_file_path)
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [10]:
.isnull()?? None??

In [11]:
data_covid.isnull().any() ##เอาค่าความจริงภายในแต่ละคอลัมม์มา  or กัน

No.                      False
announce_date            False
notification_date         True
sex                       True
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

.all() เอาความจริมาภายในแต่ละคอลัมม์มา and

In [12]:
data_covid.isnull().all()

No.                      False
announce_date            False
notification_date        False
sex                      False
age                      False
nationality              False
province_of_isolation    False
province_of_onset        False
district_of_onset        False
risk                     False
dtype: bool

In [15]:
data_covid.iloc[0,0].isnull()

AttributeError: ignored

In [16]:
data_covid.iloc[0,0]

1

In [17]:
data_covid.iloc[:1,0].isnull()

0    False
Name: No., dtype: bool

## ต่อตาราง  แกน ป y
-ต่อแกน y เพิ่มจำนวนข้อมูล

-ต่อแกน x เพิ่มคอลัมม์ (ราบละเอียดของข้อมูล)

## แกน y pd.concat()
 pd.concat

.groupby


save table

In [ ]:
datakk = data_covid[data_covid['province_of_onset']=='ขอนแก่น']
datakk

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[14 rows x 10 columns]

In [ ]:
dataud = data_covid[data_covid['province_of_onset']=='อุดรธานี']
dataud

No.  ...                                         risk
424    425  ...                                      สนามมวย
434    435  ...                                  สถานบันเทิง
471    472  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883    884  ...                                  สถานบันเทิง
885    886  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
941    942  ...                                  สถานบันเทิง
1059  1060  ...                คนไทยเดินทางกลับจากต่างประเทศ
2013  2014  ...                คนไทยเดินทางกลับจากต่างประเทศ
5810  5811  ...                            Cluster สมุทรสาคร
7433  7434  ...                                Cluster ระยอง

[10 rows x 10 columns]

In [ ]:
datams = data_covid[data_covid['province_of_onset']=='มหาสารคาม']
datams

No.  ...                                         risk
346      347  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                  สถานบันเทิง
6690    6691  ...                            Cluster สมุทรสาคร
10802  10803  ...                            Cluster สมุทรสาคร

[4 rows x 10 columns]

In [ ]:
datamyisan = pd.concat([datakk,dataud,datams])
datamyisan

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
424      425  ...                                        สนามมวย
434      435  ...                                    สถานบันเทิง
471      472  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883      884  ...                                    สถานบันเทิง
885      886  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
941      942  ...                                    สถานบันเทิง
1059    1060  ...                  คนไทยเดินทางกลับจากต่างประเทศ
2013    2014  ...                  คนไทยเดินทางกลับจากต่างประเทศ
5810    5811  ...                              Cluster สมุทรสาคร
7433    7434  ...                                  Cluster ระยอง
346      347  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                    สถานบันเทิง
6690    6691  ...                              Cluster สมุทรสาคร
10802  10803  ...                              Cluster สมุทรสาคร

[28 rows x 10 columns]

In [ ]:
data_province = data_covid[['No.','announce_date','province_of_onset']]
data_province

No. announce_date province_of_onset
0          1     1/12/2020     กรุงเทพมหานคร
1          2     1/17/2020     กรุงเทพมหานคร
2          3     1/22/2020            นครปฐม
3          4     1/22/2020     กรุงเทพมหานคร
4          5     1/24/2020     กรุงเทพมหานคร
...      ...           ...               ...
12648  12649     1/20/2021            ชลบุรี
12649  12650     1/20/2021             ระยอง
12650  12651     1/20/2021             ระยอง
12651  12652     1/20/2021             ระยอง
12652  12653     1/20/2021               ตาก

[12653 rows x 3 columns]

In [ ]:
data_human = data_covid[['No.','age','sex','nationality']]
data_human

No.   age   sex nationality
0          1  61.0  หญิง       China
1          2  74.0  หญิง       China
2          3  73.0  หญิง    Thailand
3          4  68.0   ชาย       China
4          5  66.0  หญิง       China
...      ...   ...   ...         ...
12648  12649  44.0  หญิง    Thailand
12649  12650  52.0  หญิง    Thailand
12650  12651  23.0  หญิง    Thailand
12651  12652  29.0  หญิง    Thailand
12652  12653  22.0  หญิง       Burma

[12653 rows x 4 columns]

## แบบง่าย รู้ว่าสองตารางเรคอร์ดตรงเท่ากัน

In [ ]:
full_table = data_human.merge(data_province)
full_table

No.   age   sex nationality announce_date province_of_onset
0          1  61.0  หญิง       China     1/12/2020     กรุงเทพมหานคร
1          2  74.0  หญิง       China     1/17/2020     กรุงเทพมหานคร
2          3  73.0  หญิง    Thailand     1/22/2020            นครปฐม
3          4  68.0   ชาย       China     1/22/2020     กรุงเทพมหานคร
4          5  66.0  หญิง       China     1/24/2020     กรุงเทพมหานคร
...      ...   ...   ...         ...           ...               ...
12648  12649  44.0  หญิง    Thailand     1/20/2021            ชลบุรี
12649  12650  52.0  หญิง    Thailand     1/20/2021             ระยอง
12650  12651  23.0  หญิง    Thailand     1/20/2021             ระยอง
12651  12652  29.0  หญิง    Thailand     1/20/2021             ระยอง
12652  12653  22.0  หญิง       Burma     1/20/2021               ตาก

[12653 rows x 6 columns]

## sort

In [ ]:
data_human2 = data_human.sort_values('age')
data_human2

No.   age   sex nationality
1987    1988  0.10   ชาย    Thailand
11497  11498  0.11   ชาย       Burma
6477    6478  0.11  หญิง    Thailand
1675    1676  0.30   ชาย       Japan
1075    1076  0.40   ชาย    Thailand
...      ...   ...   ...         ...
12635  12636   NaN   NaN    Thailand
12636  12637   NaN   NaN    Thailand
12637  12638   NaN   NaN    Thailand
12638  12639   NaN   NaN    Thailand
12639  12640   NaN   NaN    Thailand

[12653 rows x 4 columns]

In [ ]:
full_table2 = data_human2.merge(data_province)
full_table2

No.   age   sex nationality announce_date province_of_onset
0       1988  0.10   ชาย    Thailand      4/4/2020             ระยอง
1      11498  0.11   ชาย       Burma     1/16/2021         สมุทรสาคร
2       6478  0.11  หญิง    Thailand    12/30/2020          เพชรบุรี
3       1676  0.30   ชาย       Japan      4/1/2020     กรุงเทพมหานคร
4       1076  0.40   ชาย    Thailand     3/27/2020      สุราษฎร์ธานี
...      ...   ...   ...         ...           ...               ...
12648  12636   NaN   NaN    Thailand     1/20/2021         สมุทรสาคร
12649  12637   NaN   NaN    Thailand     1/20/2021         สมุทรสาคร
12650  12638   NaN   NaN    Thailand     1/20/2021         สมุทรสาคร
12651  12639   NaN   NaN    Thailand     1/20/2021         สมุทรสาคร
12652  12640   NaN   NaN    Thailand     1/20/2021         สมุทรสาคร

[12653 rows x 6 columns]

In [ ]:
data_human2_renamed = data_human2.rename(columns={'No.' : 'patientNumber'})
data_human2_renamed

patientNumber   age   sex nationality
1987            1988  0.10   ชาย    Thailand
11497          11498  0.11   ชาย       Burma
6477            6478  0.11  หญิง    Thailand
1675            1676  0.30   ชาย       Japan
1075            1076  0.40   ชาย    Thailand
...              ...   ...   ...         ...
12635          12636   NaN   NaN    Thailand
12636          12637   NaN   NaN    Thailand
12637          12638   NaN   NaN    Thailand
12638          12639   NaN   NaN    Thailand
12639          12640   NaN   NaN    Thailand

[12653 rows x 4 columns]

## x

In [ ]:
full_table3 = data_human2_renamed.merge(data_province, left_on= 'patientNumber', right_on='No.')
full_table3.head()

patientNumber   age   sex nationality    No. announce_date province_of_onset
0           1988  0.10   ชาย    Thailand   1988      4/4/2020             ระยอง
1          11498  0.11   ชาย       Burma  11498     1/16/2021         สมุทรสาคร
2           6478  0.11  หญิง    Thailand   6478    12/30/2020          เพชรบุรี
3           1676  0.30   ชาย       Japan   1676      4/1/2020     กรุงเทพมหานคร
4           1076  0.40   ชาย    Thailand   1076     3/27/2020      สุราษฎร์ธานี

In [ ]:
full_table3 = data_human2_renamed.merge(data_covid, how='left', left_on= 'age', right_on='age')
full_table3.head()

patientNumber  ...                                         risk
0           1988  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1          11498  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
2          11498  ...                            Cluster สมุทรสาคร
3           6478  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
4           6478  ...                            Cluster สมุทรสาคร

[5 rows x 14 columns]

คถณสมบัติของ pandas ก็คือเราสามารถสร้างคอลัมม์ ใหม่ให้ตาราง df ได้โดย 

df['=ชื่อ colunm ใหม่' ] = (list ที่มีจำนวนสมาชิกเท่าจำนวน record ของ df)

In [ ]:
data_human2_renamed['num'] = range(data_human2.shape[0])
data_human2_renamed

patientNumber   age  ...                                       detail    num
1987            1988  0.10  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้      0
11497          11498  0.11  ...                            Cluster สมุทรสาคร      1
6477            6478  0.11  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้      2
1675            1676  0.30  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้      3
1075            1076  0.40  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้      4
...              ...   ...  ...                                          ...    ...
12635          12636   NaN  ...                            Cluster สมุทรสาคร  12648
12636          12637   NaN  ...                            Cluster สมุทรสาคร  12649
12637          12638   NaN  ...                            Cluster สมุทรสาคร  12650
12638          12639   NaN  ...                            Cluster สมุทรสาคร  12651
12639          12640   NaN  ...                            Cluster สมุทรสาคร  12652

[12653 rows x 6 columns]

In [ ]:
data_human2_renamed['detail'] = data_human2_renamed['patientNumber'].map(data_covid.set_index('No.')['risk'])
data_human2_renamed

patientNumber  ...                                       detail
1987            1988  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11497          11498  ...                            Cluster สมุทรสาคร
6477            6478  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1675            1676  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1075            1076  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
...              ...  ...                                          ...
12635          12636  ...                            Cluster สมุทรสาคร
12636          12637  ...                            Cluster สมุทรสาคร
12637          12638  ...                            Cluster สมุทรสาคร
12638          12639  ...                            Cluster สมุทรสาคร
12639          12640  ...                            Cluster สมุทรสาคร

[12653 rows x 5 columns]

## map( ) เลือกมาเฉพาะคอลัมม์มาแปะ

In [ ]:
data

# HW7 สร้าตารางใหม่ ที่ค่าใน  sex เป็น missimg ทั้งหมด
-สรุปว่าทำไม record นั้นๆถึง missing 
- create pandas table 
- ต่อตาราง x y 
-.groupby
-save table

In [57]:
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

In [24]:
data_covid.shape

(12653, 10)

In [147]:
data_covid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12653 entries, 0 to 12652
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   No.                    12653 non-null  int64  
 1   announce_date          12653 non-null  object 
 2   notification_date      9858 non-null   object 
 3   sex                    10151 non-null  object 
 4   age                    8881 non-null   float64
 5   nationality            11442 non-null  object 
 6   province_of_isolation  12638 non-null  object 
 7   province_of_onset      10662 non-null  object 
 8   district_of_onset      9269 non-null   object 
 9   risk                   12303 non-null  object 
dtypes: float64(1), int64(1), object(8)
memory usage: 988.6+ KB


In [58]:
data_covid.isnull().sum()

No.                         0
announce_date               0
notification_date        2795
sex                      2502
age                      3772
nationality              1211
province_of_isolation      15
province_of_onset        1991
district_of_onset        3384
risk                      350
dtype: int64

In [202]:
data_covid1 = data_covid[data_covid['sex'].isnull()]
dt = data_covid1.fillna({
    'sex' : 'missing',
    })
dt



No. announce_date  ... district_of_onset               risk
4391    4392    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4392    4393    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4393    4394    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4394    4395    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4395    4396    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
...      ...           ...  ...               ...                ...
12635  12636     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12636  12637     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12637  12638     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12638  12639     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12639  12640     1/20/2021  ...             เมือง  Cluster สมุทรสาคร

[2502 rows x 10 columns]

In [203]:
dt1 = dt.iloc[:,1]
dt5 = dt.iloc[:,5]
dt6 = dt.iloc[:,6]
dt7 = dt.iloc[:,7]


In [204]:
import statistics as ss

In [211]:
 ss.mode(dt1)

'12/20/2020'

In [212]:
ss.mode(dt5)

'Burma'

In [214]:
ss.mode(dt6)

'สมุทรสาคร'

In [199]:
ss.mode(dt7)

'สมุทรสาคร'

In [215]:
dt[dt['announce_date']=='12/20/2020'].shape[0]

516

In [216]:
dt[dt['nationality']=='Burma'].shape[0]

1366

In [217]:
dt[dt['province_of_isolation']=='สมุทรสาคร'].shape[0]

2313

In [218]:
dt[dt['province_of_onset']=='สมุทรสาคร'].shape[0]

1358

จะพบค่า missing ของเพศมากที่สุดในคอลัมม์ province_of_isolation โดยพบมากที่สุดเป็นจังหวัดสมุทรสาคร 2313 ราย ทั้งนี้ค่า missing ที่พบอาจเกิดจากการแยกตัวไปรักษา อาจเป็นผู้ป่วยจากพื้นที่อื่น ประเทศอื่น หรือมีผู้ป่วยมากเกินไปอาจทำให้ข้อมูลตกหล่น จึงทำให้ไม่มีข้อมูลผู้ป่วย